### DISCUSSION TITLE
Better way to render Repeats

### DISCUSSION THREAD

**Discussion by charlesneimog** [DigiScore/neoscore#94] [Q&A] [answered]
Hi, there is a better way to render repeated barlines?

I am using this, and the result is perfect, but it seems wrong...

``` python

from neoscore.core.units import ZERO, Mm
from neoscore.core import neoscore
from neoscore.common import *
from neoscore.western.chordrest import NoteheadTable

neoscore.setup()
POSITION = (Mm(0), Mm(0))
staff = Staff(POSITION, None, Mm(80))
Barline(Mm(80), staff.group, barline_style.END)
Clef(ZERO, staff, 'treble')
noteheads = NoteheadTable("repeatDot", "repeatDot", "repeatDot", "repeatDot")
note = [('a', '', 4)]
Chordrest(Mm(76.5), staff, note, (int(1), int(1)), table=noteheads)
note = [('c', '', 5)]
Chordrest(Mm(76.5), staff, note, (int(1), int(1)), table=noteheads)
neoscore.show(display_page_geometry=False)

```

Thanks

**Comment by Xavman42** [comment, marked as answer]
You could alternatively use the SMuFL "repeatRight" glyph for this, though the result doesn't look as good as yours - at least not without some tweaking.

```
from neoscore.core import neoscore
from neoscore.core.music_text import MusicText
from neoscore.core.units import Mm
from neoscore.western.staff import Staff

neoscore.setup()
POSITION = (Mm(0), Mm(0))
staff = Staff(POSITION, None, Mm(80))
MusicText((staff.unit(44.5), staff.unit(4)), staff, "repeatRight")
neoscore.show(display_page_geometry=False)
```

**Reply by charlesneimog** [reply]
Thanks!

**Reply by ajyoon** [reply]
this solution is better, but note SMuFL's guidance that things like barlines should be drawn natively. Precomposed compound glyphs are typically for narrower text-based applications like using glyphs inline with performance notes for illustration.

> Scoring programs should draw their own repeat barlines using primitives to draw the thick and thin lines and repeatDots to draw the dots, not use the precomposed glyphs repeatLeft or repeatRight.

https://w3c.github.io/smufl/latest/tables/repeats.html

**Reply by ajyoon** [reply]
[see the example / screenshot i posted here](https://github.com/DigiScore/neoscore/discussions/94#discussioncomment-4730448)

**Comment by ajyoon** [comment]
As you've seen, there isn't out-of-the-box support for this right now in neoscore. But [I took a crack at an implementation here](https://github.com/DigiScore/neoscore/blob/main/examples/repeat_barline.py) which hopefully should be good for your purposes. I think the rhythm dot positioning is still not quite perfect, so feel free to tweak the example and let me know if you have a better solution.

![repeat-barline](https://user-images.githubusercontent.com/15976214/213525393-6441aac1-3168-4b51-aebb-b9e40dbb9f58.png)

**Reply by ajyoon** [reply]
also, note that you generally shouldn't be using chordrest and notehead tables to place SMuFL glyphs. Use MusicText directly for this kind of thing, as shown in the example.

**Comment by charlesneimog** [comment]
@ajyoon thanks for the example code and the function. I will use it!!!
